Logic/igtlioConnector.cxx

```diff
@@ -341,209 +310,220 @@ void igtlioConnector::RequestPushOutgoingMessages()
-int igtlioConnector::WaitForConnection()
+void* igtlioConnector::ConnectionThreadFunction(void* ptr)
 {
-  if (this->Type == TYPE_CLIENT)
+  vtkMultiThreader::ThreadInfo* vinfo = static_cast<vtkMultiThreader::ThreadInfo*>(ptr);
+  igtlioConnector* connector = static_cast<igtlioConnector*>(vinfo->UserData);
+
+  if (connector->Type == TYPE_SERVER)
     {
-    igtlioLockGuard<vtkMutexLock> lock(this->Mutex);
-    this->Sockets.push_back(Client(0, igtl::ClientSocket::New()));
+    connector->ServerSocket = igtl::ServerSocket::New();
+    if (connector->ServerSocket->CreateServer(connector->ServerPort) == -1)
+      {
+      vtkErrorWithObjectMacro(connector, "Failed to create server socket !");
+      connector->ServerStopFlag = true;
+      }
     }
 
-  while (!this->ServerStopFlag)
+  while (!connector->ServerStopFlag)
     {
-    if (this->Type == TYPE_SERVER)
+    if (connector->Type == TYPE_SERVER)
       {
-      igtl::ClientSocket::Pointer client = this->ServerSocket->WaitForConnection(1000);
+      igtl::ClientSocket::Pointer client = connector->ServerSocket->WaitForConnection(1000);
       if (client.IsNotNull()) // if client connected
         {
-        igtlioLockGuard<vtkMutexLock> lock(this->Mutex);
-        client->SetTimeout(SOCKET_TIMEOUT_MILLISECONDS);
-        this->Sockets.push_back(Client(this->NextClientID++, client));
-        return 1;
+        igtlioLockGuard<vtkMutexLock> lock(connector->Mutex);
+        int clientThreadID = connector->Thread->SpawnThread((vtkThreadFunctionType)&igtlioConnector::ReceiverThreadFunction, connector);
+        Client clientInfo = Client(connector->NextClientID++, client, clientThreadID);
+        connector->Sockets.push_back(clientInfo);
         }
       }
-    else if (this->Type == TYPE_CLIENT) // if this->Type == TYPE_CLIENT
+    else if (connector->Type == TYPE_CLIENT) // if this->Type == TYPE_CLIENT
       {
-      igtlioLockGuard<vtkMutexLock> lock(this->Mutex);
-      assert(this->Sockets.size() > 0);
-      int r = this->Sockets[0].Socket->ConnectToServer(this->ServerHostname.c_str(), this->ServerPort);
-      if (r == 0) // if connected to server
+      if (connector->Sockets.empty())
         {
-        this->Sockets[0].Socket->SetTimeout(SOCKET_TIMEOUT_MILLISECONDS);
-        return 1;
+          igtl::ClientSocket::Pointer socket = igtl::ClientSocket::New();
+          int r = socket->ConnectToServer(connector->ServerHostname.c_str(), connector->ServerPort);
+          if (r == 0) // if connected to server
+          {
+            igtlioLockGuard<vtkMutexLock> lock(connector->Mutex);
+            int clientThreadID = connector->Thread->SpawnThread((vtkThreadFunctionType)&igtlioConnector::ReceiverThreadFunction, connector);
+            Client clientInfo = Client(0, socket, clientThreadID);
+            connector->Sockets.push_back(clientInfo);
+            connector->RequestInvokeEvent(ClientConnectedEvent);
+          }
         }
-      else
+      igtl::Sleep(100);
```

Comment:
Is there any specific reason to add this? If yes, then describe it in a comment.
If it is not necessary then consider removing it.